src/main/java/graphql/util/DependencyGraph.java

```diff
@@ -0,0 +1,294 @@
+                closed.add(node);
+                unclosed.remove(node);
+                lastClosure.remove(node);
+                
+                node.fireResolved(context);
+                return true;
+            }
+            
+            return false;
+        }
+
+        private boolean closeResolved (N maybeNode) {
+            return closeNode(maybeNode, false/*autoClose*/);
+        }
+        
+        private boolean autoClose (N maybeNode) {
+            return closeNode(maybeNode, true/*autoResolve*/);
+        }
+        
+        @Override
+        public TraversalControl enter(TraverserContext<N> context) {
+            TraverserContext<N> parentContext = context.getParentContext();
+            Collection<N> closure = parentContext.getVar(Collection.class);
+            context
+                .setVar(Collection.class, closure)    // to be propagated to children
+                .setAccumulate(closure);                  // to be returned
+            
+            N node = context.thisNode();
+            if (parentContext.thisNode() == null || closed.containsAll(node.dependencySet())) {
+                if (autoClose(node)) {
+                    return TraversalControl.CONTINUE;
+                } else {
+                    closure.add(node);
+                    return TraversalControl.ABORT;
+                }                           
+            } else {
+                context
+                    .visitedNodes()
+                    .remove(node);
+                return TraversalControl.ABORT;
+            }
+        }
+
+        @Override
+        public TraversalControl leave(TraverserContext<N> context) {
+            return TraversalControl.CONTINUE;
+        }
+
+        @Override
+        public TraversalControl backRef(TraverserContext<N> context) {
+            assertShouldNeverHappen("cycle around node with id={}", context.thisNode().getId());
+            return TraversalControl.QUIT;
+        }
+        
+        protected DependenciesIteratorImpl (DependencyGraphContext context, Supplier<Set<N>> closureCreator) {
+            this.context = assertNotNull(context);
+            this.closureCreator = assertNotNull(closureCreator);
+            this.unclosed = closureCreator.get();
+            this.closed = closureCreator.get();
+            
+            unclosed.addAll(vertices.values());
+        }
+
+        protected DependenciesIteratorImpl (DependencyGraphContext context) {
+            this(context, () -> newSetFromMap(new IdentityHashMap<>()));
+        }
+        
+        final DependencyGraphContext context;
+        final Supplier<Set<N>> closureCreator;
+        final Collection<N> unclosed;
+        final Collection<N> closed;
+        final Traverser<N> traverser = Traverser.<N>breadthFirst(Vertex::adjacencySet, null);
+        Set<N> currentClosure = Collections.emptySet();
+        Set<N> lastClosure = Collections.emptySet();
+    }
+    
+    protected int nextId = 0;
```

Comment:
Suggesting to move **nextId**, **vertices**, **verticesById**, **edges** to the top as per[ java conventions](https://www.oracle.com/technetwork/java/codeconventions-150003.pdf) - 6.2 Placement. 

Line #238 is very unexpected place for key properties which defines this class as a graph. 

To maintain readability for **edges**, I would suggest to move anonymous set  into inner utility class and just initialize the variable.